builds.py

```diff
@@ -0,0 +1,313 @@
+    if items[2] == '(invalid date)':
+        date = None
+    else:
+        date = datetime.strptime(items[2], '%Y-%m-%dT%H:%M:%S.%f%z').date()
+    data = items[4]
+    return date, data
+
+def times_within(time1, time2, gap):
+    """Checks whether the gap between two times is less than gap"""
+    if time1 is None or time2 is None:
+        return None
+    return abs(time1 - time2) < gap
+
+def save_devices_to_file():
+    """Save a set of analysed devices to a JSON file"""
+    global _devices
+    output = dict()
+    for device in _devices:
+        output[str(device)] = device.dict_export()
+    with open(JSON_PATH, 'w') as f:
+        json.dump(output, f, indent=2, default=str)
+
+def get_devices_from_file():
+    """Load pre-analysed device data from a JSON file"""
+    global _devices
+    _devices = []
+    with open(JSON_PATH, 'r') as f:
+        rjson = json.load(f)
+        for device in rjson.values():
+            _devices.append(Device.dict_import(device))
+
+def _get_devices():
+    """Load devices in from analysing Device Analyzer data"""
+    global _devices
+    if _devices is None:
+        #_devices = []
+        get_devices_from_file()
+    start = len(_devices)
+    if start < MAXLEN or MAXLEN == -1:
+        dumps = os.listdir(PATH)
+        if MAXLEN == -1:
+            dumps = dumps[start:]
+        else:
+            dumps = dumps[start:MAXLEN]
+        
+        for index, dump in enumerate(dumps):
+            try:
+                with gzip.open(PATH+dump, 'rt') as infile:
+                    device = Device(dump)
+                    content = infile.readline()
+                    os_string = ''
+                    os_date = datetime(1970, 1, 1).date()
+                    build_string = ''
+                    build_date = datetime(1970, 1, 1).date()
+                    os_string_found = False
+                    while content:
+                        if ';system|osstring;' in content:
+                            if os_string_found and os_date != None:
+                                device.add_record(os_date, os_string, None)
+                            os_date, os_string = parse_line(content)
+                            os_string_found = True
+                        elif ';system|build|fingerprint;' in content:
+                            build_date, build_string = parse_line(content)
+                            if times_within(build_date, os_date, TIME_GAP):
+                                device.add_record(build_date, os_string, build_string)
+                                os_string_found = False
+                        content = infile.readline()
+                    pprint.pprint(device.release_first_seen)
+                    _devices.append(device)
+            except EOFError:
+                print('Invalid file:', dump)
+                continue
+        save_devices_to_file()
+    return _devices
+
+def analyse_vulnerabilities(vulnerabilities, dates, string_keys=False):
```

Comment:
What is string_keys for?

Reply:
The JSON export library doesn't allow datetimes to be dictionary keys, so there has to be an option to represent the dates as strings for JSON export